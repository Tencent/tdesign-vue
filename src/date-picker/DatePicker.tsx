import { defineComponent, watch, computed } from '@vue/composition-api';
import dayjs from 'dayjs';
import { CalendarIcon as TdCalendarIcon } from 'tdesign-icons-vue';

import { usePrefixClass } from '../hooks/useConfig';
import { useGlobalIcon } from '../hooks/useGlobalIcon';
import useSingle from './hooks/useSingle';
import {
  parseToDayjs, getDefaultFormat, formatTime, formatDate,
} from './hooks/useFormat';
import { subtractMonth, addMonth, extractTimeObj } from '../_common/js/date-picker/utils';
import type { DateValue } from './type';
import props from './props';

import TSelectInput from '../select-input';
import TSinglePanel from './panel/SinglePanel';

export default defineComponent({
  name: 'TDatePicker',
  props,
  setup(props, { emit }) {
    const COMPONENT_NAME = usePrefixClass('date-picker');
    const { CalendarIcon } = useGlobalIcon({ CalendarIcon: TdCalendarIcon });

    const {
      inputValue,
      popupVisible,
      inputProps: datePickerInputProps,
      popupProps: datePickerPopupProps,
      isHoverCell,
      cacheValue,
      value,
      year,
      month,
      time,
      inputRef,
      onChange,
    } = useSingle(props, { emit });

    const formatRef = computed(() => getDefaultFormat({
      mode: props.mode,
      format: props.format,
      valueType: props.valueType,
      enableTimePicker: props.enableTimePicker,
    }));

    watch(popupVisible, (visible) => {
      // 面板展开重置数据
      if (visible) {
        year.value = parseToDayjs(value.value || new Date(), formatRef.value.format).year();
        month.value = parseToDayjs(value.value || new Date(), formatRef.value.format).month();
        time.value = formatTime(value.value || new Date(), formatRef.value.timeFormat);
        if (value.value) {
          cacheValue.value = formatDate(value.value, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.format,
          });
        }
      }
    });

    // 日期 hover
    function onCellMouseEnter(date: Date) {
      isHoverCell.value = true;
      inputValue.value = formatDate(date, {
        format: formatRef.value.format,
        targetFormat: formatRef.value.format,
      });
    }

    // 日期 leave
    function onCellMouseLeave() {
      isHoverCell.value = false;
      inputValue.value = formatDate(cacheValue.value, {
        format: formatRef.value.format,
        targetFormat: formatRef.value.format,
      });
    }

    // 日期点击
    function onCellClick(date: Date) {
      isHoverCell.value = false;
      // date 模式自动切换年月
      if (props.mode === 'date') {
        year.value = date.getFullYear();
        month.value = date.getMonth();
      }
      if (props.enableTimePicker) {
        cacheValue.value = formatDate(date, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.format,
        });
      } else {
        onChange?.(
          formatDate(date, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
          }) as DateValue,
          {
            dayjsValue: dayjs(date),
            trigger: 'pick',
          },
        );
        popupVisible.value = false;
      }

      props.onPick?.(date);
      emit('pick', date);
    }

    // 头部快速切换
    function onJumperClick({ trigger }: { trigger: string }) {
      const monthCountMap = {
        date: 1,
        week: 1,
        month: 12,
        quarter: 12,
        year: 120,
      };
      const monthCount = monthCountMap[props.mode] || 0;

      const current = new Date(year.value, month.value);

      let next = null;
      if (trigger === 'prev') {
        next = subtractMonth(current, monthCount);
      } else if (trigger === 'current') {
        next = new Date();
      } else if (trigger === 'next') {
        next = addMonth(current, monthCount);
      }

      const nextYear = next.getFullYear();
      const nextMonth = next.getMonth();

      year.value = nextYear;
      month.value = nextMonth;
    }

    // timePicker 点击
    function onTimePickerChange(val: string) {
      time.value = val;

      const {
        hours, minutes, seconds, milliseconds, meridiem,
      } = extractTimeObj(val);

      // am pm 12小时制转化 24小时制
      let nextHours = hours;
      if (/am/i.test(meridiem) && nextHours === 12) nextHours -= 12;
      if (/pm/i.test(meridiem) && nextHours < 12) nextHours += 12;
      const currentDate = !dayjs(inputValue.value as string, formatRef.value.format).isValid()
        ? dayjs()
        : dayjs(inputValue.value as string, formatRef.value.format);
      const nextDate = currentDate.hour(nextHours).minute(minutes).second(seconds).millisecond(milliseconds)
        .toDate();
      inputValue.value = formatDate(nextDate, {
        format: formatRef.value.format,
        targetFormat: formatRef.value.format,
      });

      props.onPick?.(nextDate);
    }

    // 确定
    function onConfirmClick() {
      const nextValue = formatDate(inputValue.value, {
        format: formatRef.value.format,
        targetFormat: formatRef.value.format,
      });
      if (nextValue) {
        onChange?.(
          formatDate(inputValue.value, {
            format: formatRef.value.format,
            targetFormat: formatRef.value.valueType,
          }) as DateValue,
          {
            dayjsValue: dayjs(inputValue.value as string),
            trigger: 'confirm',
          },
        );
      } else {
        inputValue.value = formatDate(value.value, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.format,
        });
      }
      popupVisible.value = false;
    }

    // 预设
    function onPresetClick(presetValue: DateValue | (() => DateValue)) {
      const presetVal = typeof presetValue === 'function' ? presetValue() : presetValue;
      onChange?.(
        formatDate(presetVal, {
          format: formatRef.value.format,
          targetFormat: formatRef.value.valueType,
        }) as DateValue,
        {
          dayjsValue: dayjs(presetVal),
          trigger: 'preset',
        },
      );
      popupVisible.value = false;
    }

    function onYearChange(nextYear: number) {
      year.value = nextYear;
    }

    function onMonthChange(nextMonth: number) {
      month.value = nextMonth;
    }

    const panelProps: any = computed(() => ({
      value: cacheValue.value as string,
      year: year.value,
      month: month.value,
      format: formatRef.value.format,
      mode: props.mode,
      presets: props.presets,
      time: time.value as string,
      disableDate: props.disableDate,
      firstDayOfWeek: props.firstDayOfWeek,
      timePickerProps: props.timePickerProps,
      enableTimePicker: props.enableTimePicker,
      presetsPlacement: props.presetsPlacement,
      popupVisible: popupVisible.value,
      onCellClick,
      onCellMouseEnter,
      onCellMouseLeave,
      onJumperClick,
      onConfirmClick,
      onPresetClick,
      onYearChange,
      onMonthChange,
      onTimePickerChange,
      onPanelClick: () => inputRef.value?.focus?.(),
    }));

    return {
      COMPONENT_NAME,
      inputValue,
      datePickerPopupProps,
      datePickerInputProps,
      popupVisible,
      panelProps,
      CalendarIcon,
    };
  },
  render() {
    const {
      COMPONENT_NAME,
      inputValue,
      datePickerPopupProps,
      datePickerInputProps,
      popupVisible,
      panelProps,
      CalendarIcon,
    } = this;

    const renderSuffixIcon = () => {
      if (this.suffixIcon) return this.suffixIcon;
      if (this.$scopedSlots.suffixIcon) return this.$scopedSlots.suffixIcon;
      if (this.$scopedSlots['suffix-icon']) return this.$scopedSlots['suffix-icon'];

      return () => <CalendarIcon />;
    };

    return (
      <div class={COMPONENT_NAME}>
        <TSelectInput
          disabled={this.disabled}
          value={inputValue}
          status={this.status}
          tips={this.tips}
          popupProps={datePickerPopupProps}
          inputProps={{ suffixIcon: renderSuffixIcon(), ...datePickerInputProps }}
          popupVisible={popupVisible}
          clearable={this.clearable}
          allowInput={this.allowInput}
          panel={() => <TSinglePanel {...{ props: panelProps }} />}
        />
      </div>
    );
  },
});
