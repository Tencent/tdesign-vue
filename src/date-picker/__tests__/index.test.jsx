import { mount } from '@vue/test-utils';
import MockDate from 'mockdate';
import { nextTick } from 'vue';
import DatePicker from '@/src/date-picker/index.ts';

// 固定时间，当使用 new Date() 时，返回固定时间，防止“当前时间”的副作用影响，导致 snapshot 变更，mockdate 插件见 https://github.com/boblauer/MockDate
MockDate.set('2020-12-28');

// every component needs four parts: props/events/slots/functions.
describe('DatePicker', () => {
  // test props api
  describe(':props', () => {
    it('', () => {
      const wrapper = mount({
        render() {
          return <DatePicker />;
        },
      });
      expect(wrapper.exists()).toBe(true);
    });
  });

  it(':value', () => {
    const wrapper = mount({
      render() {
        return <DatePicker value={'1998-11-11'} />;
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it(':mode', () => {
    const wrapper = mount({
      render() {
        return <DatePicker mode={'year'} />;
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it(':range', () => {
    const wrapper = mount({
      render() {
        const testRange = [
          '2018-08', // new Date(2017, 7)
          '2028-04', // new Date(2027, 3)
        ];
        return <DatePicker value={testRange} range />;
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('clearable', async () => {
    const wrapper = mount({
      render() {
        return <DatePicker defaultValue={'2022-09-14'} clearable></DatePicker>;
      },
    });
    const input = wrapper.find('.t-input');
    await input.trigger('mouseenter');
    await nextTick();
    expect(wrapper.find('.t-input__suffix-clear').exists()).toBeTruthy();
  });

  it('disabled', () => {
    const wrapper = mount({
      render() {
        return <DatePicker disabled />;
      },
    });
    const inputInner = wrapper.find('.t-input__inner');
    expect(inputInner.element.disabled).toBe(true);
  });

  it('enableTimePicker', async () => {
    const wrapper = mount({
      render() {
        return <DatePicker enableTimePicker />;
      },
    });
    wrapper.find('.t-input').trigger('click');
    await nextTick();
    expect(document.querySelector('.t-date-picker__panel-time')).not.toBe(null);
  });

  it('firstDayOfWeek', async () => {
    const wrapper = mount(DatePicker, {
      propsData: {
        firstDayOfWeek: 3,
      },
    });
    wrapper.find('.t-input').trigger('click');
    await nextTick();
    const weekElement = document.querySelector('.t-date-picker__table table thead tr th');
    expect(weekElement.innerHTML).toEqual('一');
  });

  it('format', async () => {
    // defaultValue
    const wrapper = mount({
      render() {
        return <DatePicker defaultValue={'2022-09-14 16:30:20'} format={'YYYY-MM-DD'} />;
      },
    });
    const inputElement1 = wrapper.find('.t-input__inner');
    expect(inputElement1.element.value).toEqual('2022-09-14');

    // value
    const wrapper2 = mount({
      render() {
        return <DatePicker value={'2022-09-14 16:30:20'} format={'YYYY-MM-DD'} />;
      },
    });
    const inputElement2 = wrapper2.find('.t-input__inner');
    expect(inputElement2.element.value).toEqual('2022-09-14');
  });

  it('inputProps', () => {
    const wrapper = mount(DatePicker, {
      propsData: {
        inputProps: { inputClass: 'test-inputClass' },
      },
    });
    expect(wrapper.find('.t-input').classes()).toContain('test-inputClass');
  });

  it('mode week', async () => {
    const wrapper = mount({
      render() {
        return <DatePicker mode={'week'} value={'2022-37th'} />;
      },
    });
    wrapper.find('.t-input').trigger('click');
    await nextTick();
    expect(document.querySelector('.t-date-picker__panel-week')).not.toBe(null);
  });

  // test slots
  describe('<slot>', () => {
    it('', () => {
      window.console.info('DatePicker test<slot>');
    });
  });

  // test exposure function
  describe('function', () => {
    it('', () => {
      window.console.info('DatePicker test function');
    });
  });
});
