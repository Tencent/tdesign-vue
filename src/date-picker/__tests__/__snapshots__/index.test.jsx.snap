// Vitest Snapshot v1

exports[`DatePicker > :mode 1`] = `
<div
  class="t-date-picker"
>
  <div
    class="t-select-input t-select-input--empty"
  >
    <div
      class="t-input__wrap"
    >
      <div
        class="t-input t-is-readonly t-input--prefix t-input--suffix"
      >
        <div
          class="t-input__prefix"
        />
        <input
          autocomplete=""
          class="t-input__inner"
          placeholder="请选择年份"
          readonly="readonly"
          type="text"
          unselectable="on"
        />
        <span
          class="t-input__suffix t-input__suffix-icon"
        >
          <svg
            class="t-icon t-icon-calendar"
            fill="none"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M6 4V1.5H8V4H16V1.5H18V4H22V22H2V4H6ZM4 6V9H20V6H4ZM20 11H4V20H20V11Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DatePicker > :range 1`] = `
<div
  class="t-date-picker"
  range="true"
>
  <div
    class="t-select-input"
  >
    <div
      class="t-input__wrap"
    >
      <div
        class="t-input t-is-readonly t-input--prefix t-input--suffix"
      >
        <div
          class="t-input__prefix"
        />
        <input
          autocomplete=""
          class="t-input__inner"
          placeholder="请选择日期"
          readonly="readonly"
          type="text"
          unselectable="on"
        />
        <span
          class="t-input__suffix t-input__suffix-icon"
        >
          <svg
            class="t-icon t-icon-calendar"
            fill="none"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M6 4V1.5H8V4H16V1.5H18V4H22V22H2V4H6ZM4 6V9H20V6H4ZM20 11H4V20H20V11Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DatePicker > :value 1`] = `
<div
  class="t-date-picker"
>
  <div
    class="t-select-input"
  >
    <div
      class="t-input__wrap"
    >
      <div
        class="t-input t-is-readonly t-input--prefix t-input--suffix"
      >
        <div
          class="t-input__prefix"
        />
        <input
          autocomplete=""
          class="t-input__inner"
          placeholder="请选择日期"
          readonly="readonly"
          type="text"
          unselectable="on"
        />
        <span
          class="t-input__suffix t-input__suffix-icon"
        >
          <svg
            class="t-icon t-icon-calendar"
            fill="none"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M6 4V1.5H8V4H16V1.5H18V4H22V22H2V4H6ZM4 6V9H20V6H4ZM20 11H4V20H20V11Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
